### PR TITLE
Divide & Conquer + Jacobi (part 2)

### PR DESCRIPTION
**Pull request by jzuniga-amd**
This PR completes the divide&conquer + Jacobi methods (SYEVDJ/SYGVDJ).
It required substantial modifications to STEDC that, on the bright side, will ease further optimizations to the, now independent, divide & conquer phases.

### CODE REVIEW

File: clients/CMakeLists.txt [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # ##########################################################################
-# Copyright (C) 2019-2023 Advanced Micro Devices, Inc. All rights reserved.
+# Copyright (c) 2019-2023 Advanced Micro Devices, Inc. All rights reserved.
```
Comment (cgmb) [inline clients/CMakeLists.txt]: You may need to run `./scripts/install-hooks` again to update your hooks to the new template.

File: library/include/rocsolver/rocsolver-functions.h [hunk 2]
```diff
@@ -16484,6 +16484,1042 @@ ROCSOLVER_EXPORT rocblas_status rocsolver_zheevd_strided_batched(rocblas_handle
...
+	@param[in]
+	strideD		rocblas_stride.\n
+				Stride from the start of one vector D_j to the next one D_(j+1).
+				There is no restriction for the value of strideD. Normal use case is strideD >= n.
+    @param[out]
+    info        pointer to rocblas_int. Array of batch_count integers on the GPU.\n
+                If info[j] = 0, successful exit for A_j. If info[j] = 1, the algorithm did not converge for A_j.
+	@param[in]
+	batch_count	rocblas_int. batch_count >= 0.\n
+				Number of matrices in the batch.
```
Comment (tfalders) [inline library/include/rocsolver/rocsolver-functions.h]: Indentation on strideA, strideD, and batch_count is inconsistent. Applies to all batched functions.

File: library/include/rocsolver/rocsolver-functions.h [hunk 3]
```diff
@@ -16484,6 +16484,1042 @@ ROCSOLVER_EXPORT rocblas_status rocsolver_zheevd_strided_batched(rocblas_handle
...
+	strideA		rocblas_stride.\n
+				Stride from the start of one matrix A_j to the next one A_(j+1).
+				There is no restriction for the value of strideA. Normal use case is strideA >= lda*n.
+    @param[out]
+    D           pointer to type. Array on the GPU (the size depends on the value of strideD).\n
+                The eigenvalues of A_j in increasing order.
+	@param[in]
+	strideD		rocblas_stride.\n
+				Stride from the start of one vector D_j to the next one D_(j+1).
+				There is no restriction for the value of strideW. Normal use case is strideD >= n.
```
Comment (tfalders) [inline library/include/rocsolver/rocsolver-functions.h:16782]: strideW -> strideD

File: library/include/rocsolver/rocsolver-functions.h [hunk 4]
```diff
@@ -16484,6 +16484,1042 @@ ROCSOLVER_EXPORT rocblas_status rocsolver_zheevd_strided_batched(rocblas_handle
...
+        Z^T B Z=I & \: \text{if 1st or 2nd form, or}\\
+        Z^T B^{-1} Z=I & \: \text{if 3rd form.}
+        \end{array}
+    \f]
+
+	@param[in]
+    handle      rocblas_handle.
+	@param[in]
+	itype		#rocblas_eform.\n
+				Specifies the form of the generalized eigenproblem.
```
Comment (tfalders) [inline library/include/rocsolver/rocsolver-functions.h]: Indentation on first param and on itype is inconsistent. Applies to all sygvdj functions.
The descriptions of A, B, and info also have some inconsistent indentations going on.

File: library/include/rocsolver/rocsolver-functions.h [hunk 5]
```diff
@@ -16484,6 +16484,1042 @@ ROCSOLVER_EXPORT rocblas_status rocsolver_zheevd_strided_batched(rocblas_handle
...
+    batch of real generalized symmetric-definite eigenproblems.
+
+    \details
+    For each instance in the batch, the problem solved by this function is either of the form
+
+    \f[
+        \begin{array}{cl}
+        A_j X_j = \lambda B_j X_j & \: \text{1st form,}\\
+        A_j B_j X_j = \lambda X_j & \: \text{2nd form, or}\\
+        B_j A_j X_j = \lambda X_j & \: \text{3rd form,}
```
Comment (tfalders) [inline library/include/rocsolver/rocsolver-functions.h]: Should lambda have a j subscript too?

File: library/include/rocsolver/rocsolver-functions.h [hunk 6]
```diff
@@ -16484,6 +16484,1042 @@ ROCSOLVER_EXPORT rocblas_status rocsolver_zheevd_strided_batched(rocblas_handle
...
+    @param[in]
+    ldb         rocblas_int. ldb >= n.\n
+                Specifies the leading dimension of matrices B_j.
+    @param[out]
+    D           pointer to type. Array on the GPU (the size depends on the value of strideD).\n
+                The eigenvalues in increasing order.
+	@param[in]
+	strideD		rocblas_stride.\n
+				Stride from the start of one vector D_j to the next one D_(j+1).
+				There is no restriction for the value of strideW. Normal use is strideD >= n.
```
Comment (tfalders) [inline library/include/rocsolver/rocsolver-functions.h:17145]: strideW -> strideD

File: library/include/rocsolver/rocsolver-functions.h [hunk 7]
```diff
@@ -16484,6 +16484,1042 @@ ROCSOLVER_EXPORT rocblas_status rocsolver_zheevd_strided_batched(rocblas_handle
...
+    strideB     rocblas_stride.\n
+                Stride from the start of one matrix B_j to the next one B_(j+1).
+                There is no restriction for the value of strideB. Normal use is strideB >= ldb*n.
+    @param[out]
+    D           pointer to type. Array on the GPU (the size depends on the value of strideD).\n
+                The eigenvalues in increasing order.
+	@param[in]
+	strideD		rocblas_stride.\n
+				Stride from the start of one vector D_j to the next one D_(j+1).
+				There is no restriction for the value of strideW. Normal use is strideD >= n.
```
Comment (tfalders) [inline library/include/rocsolver/rocsolver-functions.h:17258]: strideW -> strideD

File: library/include/rocsolver/rocsolver-functions.h [hunk 8]
```diff
@@ -16484,6 +16484,1042 @@ ROCSOLVER_EXPORT rocblas_status rocsolver_zheevd_strided_batched(rocblas_handle
...
+				Stride from the start of one matrix A_j to the next one A_(j+1).
+				There is no restriction for the value of strideA. Normal use is strideA >= lda*n.
+    @param[inout]
+    B           pointer to type. Array on the GPU (the size depends on the value of strideB).\n
+                On entry, the symmetric positive definite matrices B_j. On exit,
+				the triangular factor of B_j as returned by \ref rocsolver_spotrf_strided_batched "POTRF_STRIDED_BATCHED".
+    @param[in]
+    ldb         rocblas_int. ldb >= n.\n
+                Specifies the leading dimension of matrices B_j.
+	@param[in]
```
Comment (tfalders) [inline library/include/rocsolver/rocsolver-functions.h:17248]: Indentation on param for strideB is inconsistent.

File: library/include/rocsolver/rocsolver-functions.h [hunk 9]
```diff
@@ -16484,6 +16484,1042 @@ ROCSOLVER_EXPORT rocblas_status rocsolver_zheevd_strided_batched(rocblas_handle
...
+                                                          const rocblas_int ldb,
+                                                          double* D,
+                                                          const rocblas_stride strideD,
+                                                          rocblas_int* info,
+                                                          const rocblas_int batch_count);
+//! @}
+
+/*! @{
+    \brief SYGVDJ_STRIDED_BATCHED computes the eigenvalues and (optionally) eigenvectors of
+    batch of real generalized symmetric-definite eigenproblems.
```
Comment (tfalders) [inline library/include/rocsolver/rocsolver-functions.h]: Shouldn't SYGVDJ_STRIDED_BATCHED come after HEGVDJ_BATCHED?

**Review by tfalders — COMMENTED** [review]
Even though I'm on vacation, I would like to finish my review soon. All I have left is the code for stedc, but here are my comments for everything else (mainly rocsolver-functions.h).

**Review by tfalders — APPROVED** [review]
I've only skimmed over the stedc code, but it looks good.

EDIT: There seem to be test failures on CI though.

File: library/include/rocsolver/rocsolver-functions.h (hunk 5 above)
Comment (jzuniga-amd) [reply library/include/rocsolver/rocsolver-functions.h]: mmm I think rather X_j should be a "generic x" too (the computed eigenvectors are in Z_j)...
We may need to change all the other APIs for generalized eigenproblems, though, so I am going to leave it as is for now. 
Replacing 'j' by 'l' on the other hand.

**Pull request merged**